tests/Sub/SqsSnsQueueTest.php

```diff
@@ -70,124 +59,154 @@ protected function setUp(): void
-            'TopicArn:123456' => '\\TopicListener',
-        ]);
-        $queue->setContainer(new Container);
-        $job = $queue->pop();
+        $queue = new SqsSnsQueue($sqs, 'default');
+        $queue->setContainer($this->app);
+        $queue->pop();
 
-        $this->assertInstanceOf(SqsSnsJob::class, $job);
+        Event::assertDispatched('TopicArn:123456');
     }
 
     /** @test */
-    public function it_will_only_handle_rich_notifications_not_the_ones_with_raw_payloads()
+    public function it_can_dispatch_an_internal_event_for_topic_based_events()
     {
-        Log::shouldReceive('error')->once()->with(
-            m::pattern('/^SqsSnsQueue: Invalid SNS payload/'),
-            m::type('array')
-        );
-
-        $this->mockedSqsClient->method('receiveMessage')->willReturn(new \Aws\Result([
-            'Messages' => [$this->mockedRawMessageData],
+        Event::fake();
+        $sqs = m::mock(SqsClient::class);
+        $sqs->shouldReceive('receiveMessage')->once()->andReturn(new \Aws\Result([
+            'Messages' => [$this->mockedMessageDataWithTopicOnly],
         ]));
 
-        $queue = new SqsSnsQueue($this->mockedSqsClient, 'default');
-        $queue->setContainer(new Container);
+        $queue = new SqsSnsQueue($sqs, 'default');
+        $queue->setContainer($this->app);
         $queue->pop();
+
+        Event::assertDispatched('TopicArn:123456', function ($event, $payload) {
+            return $payload === [
+                'subject' => '',
+                'payload' => ['foo' => 'bar'],
+            ];
+        });
     }
 
     /** @test */
-    public function it_will_set_the_event_listeners_mapping()
+    public function it_can_dispatch_an_internal_event_for_subject_based_events()
     {
-        $queue = new SqsSnsQueue($this->mockedSqsClient, 'default', '', '', [
-            'Subject#action' => '\\SubjectListener',
-        ]);
+        Event::fake();
+        $sqs = m::mock(SqsClient::class);
+        $sqs->shouldReceive('receiveMessage')->once()->andReturn(new \Aws\Result([
+            'Messages' => [$this->mockedMessageDataWithTopicAndSubject],
+        ]));
 
-        $queueReflection = new \ReflectionClass($queue);
-        $eventsProperty = $queueReflection->getProperty('events');
-        $eventsProperty->setAccessible(true);
+        $queue = new SqsSnsQueue($sqs, 'default');
+        $queue->setContainer($this->app);
+        $queue->pop();
 
-        $this->assertEquals(['Subject#action' => '\\SubjectListener'], $eventsProperty->getValue($queue));
+        Event::assertDispatched('Subject#action', function ($event, $payload) {
+            return $payload === [
+                'subject' => 'Subject#action',
+                'payload' => ['foo' => 'bar'],
+            ];
+        });
     }
 
     /** @test */
-    public function it_will_resolve_the_listener_name_based_on_the_topic_arn()
+    public function it_dispatches_a_subject_based_events_over_topic_based()
```

Comment:
What does it mean for an event to be 'subject based' or 'topic based'? I found the previous test names a bit clearer